core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -172,21 +204,29 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
     }
 
     int lenCopy = len;
+    CountingRetry retry = new CountingRetry(MAX_WORKERS_TO_RETRY);
     while (len > 0) {
       if (pos >= mLength) {
         break;
       }
       long blockId = mStatus.getBlockIds().get(Math.toIntExact(pos / mBlockSize));
       BlockInStream stream = null;
+      stream = mBlockStore.getInStream(blockId, mOptions, mFailedWorkers);
       try {
-        stream = mBlockStore.getInStream(blockId, mOptions);
         long offset = pos % mBlockSize;
         int bytesRead =
             stream.positionedRead(offset, b, off, (int) Math.min(mBlockSize - offset, len));
         Preconditions.checkState(bytesRead > 0, "No data is read before EOF");
         pos += bytesRead;
         off += bytesRead;
         len -= bytesRead;
+        retry.reset();
+      } catch (UnavailableException | DeadlineExceededException | ConnectException e) {
+        if (!retry.attempt()) {
```

Comment:
@bf8086 Please check if this still correct.

Reply:
same here.